### PR TITLE
Reduce size by new url generation and counter reuse

### PR DESCRIPTION
**Pull request by gwer**
Url generation based on @anber_ru's proposal.

### CODE REVIEW

**Comment by Anber** [comment]
You forgot about `non-secure/index.js` and `url.js`. Also, the comment about low entropy is obsolete :)

File: async/index.browser.js [hunk 1]
```diff
@@ -7,36 +7,25 @@ var crypto = self.crypto || self.msCrypto
...
+  size = size || 21
+
   // Compact alternative for `for (var i = 0; i < size; i++)`
   while (size--) {
     // We can’t use bytes bigger than the alphabet. 63 is 00111111 bitmask.
     // This mask reduces random byte 0-255 to 0-63 values.
     // There is no need in `|| ''` and `* 1.6` hacks in here,
     // because bitmask trim bytes exact to alphabet size.
-    id += url[bytes[size] & 63]
+    id += url[crypto.getRandomValues(new Uint8Array(1)) & 63]
```
Comment (ai) [inline async/index.browser.js]: This change will kill performance since the single call of `getRandomValues` is much faster than multiple calls.

Comment (gwer) [reply async/index.browser.js]: image

File: async/index.browser.js [hunk 2]
```diff
@@ -7,36 +7,30 @@ var crypto = self.crypto || self.msCrypto
 // Despite the fact the source code is quite long, its entropy
 // is low and there are lots of duplicates - just what compressors
```
Comment (ai) [inline async/index.browser.js]: You need to update comment too

File: async/index.browser.js [hunk 3]
```diff
@@ -7,36 +7,30 @@ var crypto = self.crypto || self.msCrypto
...
-  i = 48, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1,
+var url = '-_'
 
-  // ASCII codes for A...Z
-  i += 8, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1,
+var i = 36
```
Comment (ai) [inline async/index.browser.js]: Let’s comment on what we generate here

File: async/index.browser.js [hunk 4]
```diff
@@ -7,36 +7,30 @@ var crypto = self.crypto || self.msCrypto
...
+}
 
-  // ASCII codes for a...z
-  i += 7, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1
-)
+i = 36
+while (i-- - 10) {
```
Comment (ai) [inline async/index.browser.js]: We need a comment about this `while` loop

File: async/index.browser.js [hunk 5]
```diff
@@ -7,36 +7,30 @@ var crypto = self.crypto || self.msCrypto
...
+var url = '-_'
 
-  // ASCII codes for A...Z
-  i += 8, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1,
+var i = 36
+while (i--) {
+  url += i.toString(36)
```
Comment (ai) [inline async/index.browser.js]: Let’s comment `36` argument of `toString`

Comment (gwer) [reply async/index.browser.js]: Done

File: async/index.browser.js [hunk 6]
```diff
@@ -4,39 +4,38 @@
...
-  i += 8, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1, i += 1,
-  i += 1, i += 1,
+var i = 36
+while (i--) {
+  // 36 is radix.
+  // Number.prototype.toString(36) returns number in Base36 representation
+  // Base36 is like hex,
+  // but Base36 is represented using the numerals 0–9 and the Latin letters A–Z
```
Comment (Anber) [inline async/index.browser.js]: Latin letters ***a-z***

Comment (gwer) [reply async/index.browser.js]: Вообще писал про Base36 в целом (где понятия регистра вообще нет), но, пожалуй, тут описание ближе к реализации из JS будет корректнее, да.

**Comment by ai** [comment]
Ещё нужно сравнить результаты `test/benchmark` до и после этого PR

**Comment by gwer** [comment]
```
$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

$ node test/benchmark
nanoid                    282,863 ops/sec
nanoid/generate           255,983 ops/sec
uid.sync                  170,236 ops/sec
uuid/v4                   202,241 ops/sec
secure-random-string      189,137 ops/sec
shortid                    30,185 ops/sec

Async:
nanoid/async               31,993 ops/sec
nanoid/async/generate      31,449 ops/sec
secure-random-string       30,437 ops/sec
uid                        29,401 ops/sec

Non-secure:
nanoid/non-secure       1,714,070 ops/sec
rndm                    1,593,879 ops/sec

$ git checkout -
Switched to branch 'f/reduce_size_with_random_in_loop'
Your branch is up to date with 'origin/f/reduce_size_with_random_in_loop'.

$ node test/benchmark
nanoid                    294,016 ops/sec
nanoid/generate           265,646 ops/sec
uid.sync                  201,971 ops/sec
uuid/v4                   219,655 ops/sec
secure-random-string      183,223 ops/sec
shortid                    28,493 ops/sec

Async:
nanoid/async               33,153 ops/sec
nanoid/async/generate      30,710 ops/sec
secure-random-string       28,456 ops/sec
uid                        28,692 ops/sec

Non-secure:
nanoid/non-secure       1,552,124 ops/sec
rndm                    1,683,481 ops/sec
```

**Comment by ai** [comment]
Released in 2.1.11

File: async/index.browser.js [hunk 7]
```diff
@@ -4,39 +4,38 @@
...
+while (i-- - 10) {
+  url += i.toString(36).toUpperCase()
+}
 
 module.exports = function (size) {
-  size = size || 21
   var id = ''
-  var bytes = crypto.getRandomValues(new Uint8Array(size))
+  var bytes = crypto.getRandomValues(new Uint8Array(size || 21))
+  i = size || 21
```
Comment (alfaproject) [inline async/index.browser.js]: Why not?
```js
  i = size || 21
  var bytes = crypto.getRandomValues(new Uint8Array(i))
```

Comment (gwer) [reply async/index.browser.js]: For better gzip compression. It makes +1 byte.

Comment (alfaproject) [reply async/index.browser.js]: Really? o:

Comment (gwer) [reply async/index.browser.js]: Sorry, I was wrong. +3 bytes :D

Comment (alfaproject) [reply async/index.browser.js]: Haha

**Pull request merged**